### ISSUE TITLE
Multiple servers working as a cluster

### ISSUE THREAD

**Issue by wandenberg** [wandenberg/nginx-push-stream-module#42]
Add the functionality to multiple servers works as a cluster, without use external tools like redis or memcached

**Comment by bartebor** [comment]
+1 for setting message id by an application. It makes use of custom versioning possible and ensures consistency between multiple modules in load balanced environments.

I would also really like to have an upstream mode, which would help make clustered and load balanced farm of servers where backend application does not have to know all servers locations (some form of inversion of control). We could use one instance of module as an upstream server or make backend to be one.

It would be also nice to have conditional get based on version - when you already have current version you will not get it again, even if hitting another server.
Why do we need this? Well, few years ago we have developed custom live distribution mechanism which is still in production.
It consists of a backend (HA clustered) and a farm of frontends, which connect to the backend.
The whole concept is based on events. Clients create events (some sort of channel, think of a sport event) and than start to publish data. Publishing starts with sending data to the backend. Backend compares it with previous version and, if different, increments version number, creates differential version (currently we have only diff-like method) and stores both diff and full data in persistend storage. Note that application always sends to backend full data, which is often quite big.

Now, if there are any frontends subscribed to this event, they are informed of changes.
Frontend maintains persistent connection to backend and subscribes to events as first clients connects and unsubscribe, when all are gone. Frontend is also responsible for sending data to clients in optimized way. We have implemented simple protocol, where (Javascript) client provides frontend with current data version. Frontend can now decide how to send data - if client is behind more than one version it sends it full current data, and a differential (potentially much smaller) version otherwise. If there is no new version, client waits on the connection.

As you see our system is some sort of state replicator as opposed to bare publish-subscribe.
It has also the advantage of self healing after server restart and supports events that are no longer live (so there will be no updates anymore). Such events have a special flag which makes client disconnect after it reads data since no updates are expected. This substantially lowers frontend load in terms of used memory and number of persisting network connections. 

Now we are looking for some alternatives, as after these years new technologies arrived such as websockets (we have only long polling). Nginx push module would be ideal replacement for our frontends, but in current shape it just does not fit... Having custom versions support, upstream mode and (possibly) support for specifying current version by client would be great :)

Since this issue is idle for a quite long time are there any possibilities of implementing these features?

Regards

**Comment by amonaco** [comment]
how about using a "scalability protocol" library approach like http://zeromq.org/ or https://github.com/nanomsg/nanomsg

Thanks!

**Comment by dennismartensson** [comment]
Hi, Is this still on the road map?

**Comment by wandenberg** [comment]
yes. but, unfortunately I'm currently "out of time"

**Comment by dennismartensson** [comment]
Okey, I have bin looking at http://nats.io/ as a messaging systems between websocket servers and have had good results in terms of speed and scaling there is a official nginx client https://github.com/nats-io/nginx-nats

I don't know how you have bin thinking about building the cluster solution. Just wanted to mention it.

**Comment by chudzini** [comment]
> May be I choose put the message id as an optional header, with that the application set the id, instead of an incremental, once the module does not use this number

Is this option available? Can I set message id on the application side?
